<a href="https://colab.research.google.com/github/fernandopontes398/pandas_gasolina/blob/main/pandas_gasolina.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [6]:
import pandas as pd


# 1. Carregando os conjuntos de dados "gasolina_2000+.csv" e "gasolina_2010+.csv" em dois DataFrames diferentes e combinando-os em um único DataFrame.

In [7]:
df1 = pd.read_csv('gasolina_2000+.csv', sep=",", decimal=".", index_col= 0)

In [8]:
df1


DATA INICIAL  DATA FINAL        REGIÃO              ESTADO  \
0       2004-05-09  2004-05-15  CENTRO OESTE    DISTRITO FEDERAL   
1       2004-05-09  2004-05-15  CENTRO OESTE               GOIAS   
2       2004-05-09  2004-05-15  CENTRO OESTE         MATO GROSSO   
3       2004-05-09  2004-05-15  CENTRO OESTE  MATO GROSSO DO SUL   
4       2004-05-09  2004-05-15      NORDESTE             ALAGOAS   
...            ...         ...           ...                 ...   
52802   2010-12-26  2011-01-01       SUDESTE      RIO DE JANEIRO   
52803   2010-12-26  2011-01-01       SUDESTE           SAO PAULO   
52804   2010-12-26  2011-01-01           SUL              PARANA   
52805   2010-12-26  2011-01-01           SUL   RIO GRANDE DO SUL   
52806   2010-12-26  2011-01-01           SUL      SANTA CATARINA   

                PRODUTO  NÚMERO DE POSTOS PESQUISADOS UNIDADE DE MEDIDA  \
0      ETANOL HIDRATADO                           127              R$/l   
1      ETANOL HIDRATADO                           387              R$/l   
2      ETANOL HIDRATADO                           192              R$/l   
3      ETANOL HIDRATADO                           162              R$/l   
4      ETANOL HIDRATADO                           103              R$/l   
...                 ...                           ...               ...   
52802       ÓLEO DIESEL                           574              R$/l   
52803       ÓLEO DIESEL                          1871              R$/l   
52804       ÓLEO DIESEL                           469              R$/l   
52805       ÓLEO DIESEL                           419              R$/l   
52806       ÓLEO DIESEL                           328              R$/l   

       PREÇO MÉDIO REVENDA  DESVIO PADRÃO REVENDA  PREÇO MÍNIMO REVENDA  \
0                    1.288                  0.016                 1.190   
1                    1.162                  0.114                 0.890   
2                    1.389                  0.097                 1.180   
3                    1.262                  0.070                 1.090   
4                    1.181                  0.078                 1.050   
...                    ...                    ...                   ...   
52802                1.983                  0.086                 1.779   
52803                1.966                  0.070                 1.699   
52804                1.946                  0.058                 1.799   
52805                2.052                  0.059                 1.890   
52806                2.018                  0.066                 1.849   

       PREÇO MÁXIMO REVENDA MARGEM MÉDIA REVENDA  COEF DE VARIAÇÃO REVENDA  \
0                     1.350                0.463                     0.012   
1                     1.449                0.399                     0.098   
2                     1.760                0.419                     0.070   
3                     1.509                0.432                     0.055   
4                     1.400                 0.24                     0.066   
...                     ...                  ...                       ...   
52802                 2.280                0.255                     0.043   
52803                 2.399                0.261                     0.036   
52804                 2.109                 0.19                     0.030   
52805                 2.310                0.267                     0.029   
52806                 2.210                0.225                     0.033   

      PREÇO MÉDIO DISTRIBUIÇÃO DESVIO PADRÃO DISTRIBUIÇÃO  \
0                        0.825                       0.11   
1                        0.763                      0.088   
2                         0.97                      0.095   
3                         0.83                      0.119   
4                        0.941                      0.077   
...                        ...                        ...   
52802                  

In [9]:
df2 = pd.read_csv('gasolina_2010+.csv', sep=",", decimal=".", index_col= 0,low_memory=False)

In [10]:
df2

DATA INICIAL  DATA FINAL        REGIÃO              ESTADO  \
9256     2011-01-02  2011-01-08  CENTRO OESTE    DISTRITO FEDERAL   
9257     2011-01-02  2011-01-08  CENTRO OESTE               GOIAS   
9258     2011-01-02  2011-01-08  CENTRO OESTE         MATO GROSSO   
9259     2011-01-02  2011-01-08  CENTRO OESTE  MATO GROSSO DO SUL   
9260     2011-01-02  2011-01-08      NORDESTE             ALAGOAS   
...             ...         ...           ...                 ...   
120818   2021-04-25  2021-05-01         NORTE             RORAIMA   
120819   2021-04-25  2021-05-01           SUL      SANTA CATARINA   
120820   2021-04-25  2021-05-01       SUDESTE           SAO PAULO   
120821   2021-04-25  2021-05-01      NORDESTE             SERGIPE   
120822   2021-04-25  2021-05-01         NORTE           TOCANTINS   

                 PRODUTO  NÚMERO DE POSTOS PESQUISADOS UNIDADE DE MEDIDA  \
9256    ETANOL HIDRATADO                            80              R$/l   
9257    ETANOL HIDRATADO                           255              R$/l   
9258    ETANOL HIDRATADO                           141              R$/l   
9259    ETANOL HIDRATADO                           132              R$/l   
9260    ETANOL HIDRATADO                            97              R$/l   
...                  ...                           ...               ...   
120818   OLEO DIESEL S10                            12              R$/l   
120819   OLEO DIESEL S10                           100              R$/l   
120820   OLEO DIESEL S10                           886              R$/l   
120821   OLEO DIESEL S10                            33              R$/l   
120822   OLEO DIESEL S10                            18              R$/l   

        PREÇO MÉDIO REVENDA  DESVIO PADRÃO REVENDA  PREÇO MÍNIMO REVENDA  \
9256                  2.030                  0.003                 2.020   
9257                  1.717                  0.102                 1.490   
9258                  1.848                  0.148                 1.590   
9259                  1.893                  0.108                 1.699   
9260                  2.026                  0.084                 1.749   
...                     ...                    ...                   ...   
120818                4.582                  0.056                 4.510   
120819                4.141                  0.187                 3.838   
120820                4.246                  0.216                 3.769   
120821                4.464                  0.205                 4.190   
120822                4.352                  0.156                 4.199   

        PREÇO MÁXIMO REVENDA MARGEM MÉDIA REVENDA  COEF DE VARIAÇÃO REVENDA  \
9256                   2.050                0.411                     0.001   
9257                   1.990                0.225                     0.059   
9258                   2.490                0.326                     0.080   
9259                   2.199                0.191                     0.057   
9260                   2.100                0.268                     0.041   
...                      ...                  ...                       ...   
120818                 4.690             -99999.0                     0.012   
120819                 4.739             -99999.0                     0.045   
120820                 5.199             -99999.0                     0.051   
120821                 4.699             -99999.0                     0.046   
120822                 4.699             -99999.0                     0.036   

       PREÇO MÉDIO DISTRIBUIÇÃO DESVIO PADRÃO DISTRIBUIÇÃO  \
9256                      1.619                      0.013   
9257                      1.492                      0.069   
9258                      1.522                      0.105   
9259                      1.702                       0.04   
9260                      1.758                      0.042   
...                         ...

In [11]:
df_combinado = pd.concat([df1, df2])



In [12]:
df_combinado

DATA INICIAL  DATA FINAL        REGIÃO              ESTADO  \
0        2004-05-09  2004-05-15  CENTRO OESTE    DISTRITO FEDERAL   
1        2004-05-09  2004-05-15  CENTRO OESTE               GOIAS   
2        2004-05-09  2004-05-15  CENTRO OESTE         MATO GROSSO   
3        2004-05-09  2004-05-15  CENTRO OESTE  MATO GROSSO DO SUL   
4        2004-05-09  2004-05-15      NORDESTE             ALAGOAS   
...             ...         ...           ...                 ...   
120818   2021-04-25  2021-05-01         NORTE             RORAIMA   
120819   2021-04-25  2021-05-01           SUL      SANTA CATARINA   
120820   2021-04-25  2021-05-01       SUDESTE           SAO PAULO   
120821   2021-04-25  2021-05-01      NORDESTE             SERGIPE   
120822   2021-04-25  2021-05-01         NORTE           TOCANTINS   

                 PRODUTO  NÚMERO DE POSTOS PESQUISADOS UNIDADE DE MEDIDA  \
0       ETANOL HIDRATADO                           127              R$/l   
1       ETANOL HIDRATADO                           387              R$/l   
2       ETANOL HIDRATADO                           192              R$/l   
3       ETANOL HIDRATADO                           162              R$/l   
4       ETANOL HIDRATADO                           103              R$/l   
...                  ...                           ...               ...   
120818   OLEO DIESEL S10                            12              R$/l   
120819   OLEO DIESEL S10                           100              R$/l   
120820   OLEO DIESEL S10                           886              R$/l   
120821   OLEO DIESEL S10                            33              R$/l   
120822   OLEO DIESEL S10                            18              R$/l   

        PREÇO MÉDIO REVENDA  DESVIO PADRÃO REVENDA  PREÇO MÍNIMO REVENDA  \
0                     1.288                  0.016                 1.190   
1                     1.162                  0.114                 0.890   
2                     1.389                  0.097                 1.180   
3                     1.262                  0.070                 1.090   
4                     1.181                  0.078                 1.050   
...                     ...                    ...                   ...   
120818                4.582                  0.056                 4.510   
120819                4.141                  0.187                 3.838   
120820                4.246                  0.216                 3.769   
120821                4.464                  0.205                 4.190   
120822                4.352                  0.156                 4.199   

        PREÇO MÁXIMO REVENDA MARGEM MÉDIA REVENDA  COEF DE VARIAÇÃO REVENDA  \
0                      1.350                0.463                     0.012   
1                      1.449                0.399                     0.098   
2                      1.760                0.419                     0.070   
3                      1.509                0.432                     0.055   
4                      1.400                 0.24                     0.066   
...                      ...                  ...                       ...   
120818                 4.690             -99999.0                     0.012   
120819                 4.739             -99999.0                     0.045   
120820                 5.199             -99999.0                     0.051   
120821                 4.699             -99999.0                     0.046   
120822                 4.699             -99999.0                     0.036   

       PREÇO MÉDIO DISTRIBUIÇÃO DESVIO PADRÃO DISTRIBUIÇÃO  \
0                         0.825                       0.11   
1                         0.763                      0.088   
2                          0.97                      0.095   
3                          0.83                      0.119   
4                         0.941                      0.077   
...                         ...

#  2.Entendendo o conjunto de dados usando o head() e info()

In [13]:
df_combinado.head()

DATA INICIAL  DATA FINAL        REGIÃO              ESTADO  \
0   2004-05-09  2004-05-15  CENTRO OESTE    DISTRITO FEDERAL   
1   2004-05-09  2004-05-15  CENTRO OESTE               GOIAS   
2   2004-05-09  2004-05-15  CENTRO OESTE         MATO GROSSO   
3   2004-05-09  2004-05-15  CENTRO OESTE  MATO GROSSO DO SUL   
4   2004-05-09  2004-05-15      NORDESTE             ALAGOAS   

            PRODUTO  NÚMERO DE POSTOS PESQUISADOS UNIDADE DE MEDIDA  \
0  ETANOL HIDRATADO                           127              R$/l   
1  ETANOL HIDRATADO                           387              R$/l   
2  ETANOL HIDRATADO                           192              R$/l   
3  ETANOL HIDRATADO                           162              R$/l   
4  ETANOL HIDRATADO                           103              R$/l   

   PREÇO MÉDIO REVENDA  DESVIO PADRÃO REVENDA  PREÇO MÍNIMO REVENDA  \
0                1.288                  0.016                  1.19   
1                1.162                  0.114                  0.89   
2                1.389                  0.097                  1.18   
3                1.262                  0.070                  1.09   
4                1.181                  0.078                  1.05   

   PREÇO MÁXIMO REVENDA MARGEM MÉDIA REVENDA  COEF DE VARIAÇÃO REVENDA  \
0                 1.350                0.463                     0.012   
1                 1.449                0.399                     0.098   
2                 1.760                0.419                     0.070   
3                 1.509                0.432                     0.055   
4                 1.400                 0.24                     0.066   

  PREÇO MÉDIO DISTRIBUIÇÃO DESVIO PADRÃO DISTRIBUIÇÃO  \
0                    0.825                       0.11   
1                    0.763                      0.088   
2                     0.97                      0.095   
3                     0.83                      0.119   
4                    0.941                      0.077   

  PREÇO MÍNIMO DISTRIBUIÇÃO PREÇO MÁXIMO DISTRIBUIÇÃO  \
0                    0.4201                    0.9666   
1                    0.5013                      1.05   
2                    0.5614                     1.161   
3                    0.5991                   1.22242   
4                    0.7441                    1.0317   

  COEF DE VARIAÇÃO DISTRIBUIÇÃO  
0                         0.133  
1                         0.115  
2                         0.098  
3                         0.143  
4                         0.082

In [14]:
df_combinado.info()

<class 'pandas.core.frame.DataFrame'>
Index: 120823 entries, 0 to 120822
Data columns (total 18 columns):
 #   Column                         Non-Null Count   Dtype  
---  ------                         --------------   -----  
 0   DATA INICIAL                   120823 non-null  object 
 1   DATA FINAL                     120823 non-null  object 
 2   REGIÃO                         120823 non-null  object 
 3   ESTADO                         120823 non-null  object 
 4   PRODUTO                        120823 non-null  object 
 5   NÚMERO DE POSTOS PESQUISADOS   120823 non-null  int64  
 6   UNIDADE DE MEDIDA              120823 non-null  object 
 7   PREÇO MÉDIO REVENDA            120823 non-null  float64
 8   DESVIO PADRÃO REVENDA          120823 non-null  float64
 9   PREÇO MÍNIMO REVENDA           120823 non-null  float64
 10  PREÇO MÁXIMO REVENDA           120823 non-null  float64
 11  MARGEM MÉDIA REVENDA           120823 non-null  object 
 12  COEF DE VARIAÇÃO REVENDA       1208

#3. Verificando o tipo da coluna DATA INICIAL.

In [15]:
df_combinado['DATA INICIAL'][2]

'2004-05-09'

# 4.As colunas DATA INICIAL e DATA FINAL estão formatadas como string. Utilizando o método pd.to_datetime(), para converter ambas colunas para Timestamp / Datetime.

In [16]:
df_combinado['DATA INICIAL'] = pd.to_datetime(df_combinado['DATA INICIAL'], errors='coerce')
df_combinado['DATA FINAL'] = pd.to_datetime(df_combinado['DATA FINAL'], errors='coerce')

df_combinado.info()

<class 'pandas.core.frame.DataFrame'>
Index: 120823 entries, 0 to 120822
Data columns (total 18 columns):
 #   Column                         Non-Null Count   Dtype         
---  ------                         --------------   -----         
 0   DATA INICIAL                   120823 non-null  datetime64[ns]
 1   DATA FINAL                     120823 non-null  datetime64[ns]
 2   REGIÃO                         120823 non-null  object        
 3   ESTADO                         120823 non-null  object        
 4   PRODUTO                        120823 non-null  object        
 5   NÚMERO DE POSTOS PESQUISADOS   120823 non-null  int64         
 6   UNIDADE DE MEDIDA              120823 non-null  object        
 7   PREÇO MÉDIO REVENDA            120823 non-null  float64       
 8   DESVIO PADRÃO REVENDA          120823 non-null  float64       
 9   PREÇO MÍNIMO REVENDA           120823 non-null  float64       
 10  PREÇO MÁXIMO REVENDA           120823 non-null  float64       
 11  MARGE

#  5. Criando uma nova coluna para representar o ano e o mês(aaaa-mm), utilizando a coluna DATA FINAL como referência.

In [17]:
df_combinado["ANO-MES"] = df_combinado["DATA FINAL"].apply(lambda x: "{}".format(x.year)) + df_combinado["DATA FINAL"].apply(lambda x: "/{:02d}".format(x.month))

In [18]:
df_combinado

DATA INICIAL DATA FINAL        REGIÃO              ESTADO  \
0        2004-05-09 2004-05-15  CENTRO OESTE    DISTRITO FEDERAL   
1        2004-05-09 2004-05-15  CENTRO OESTE               GOIAS   
2        2004-05-09 2004-05-15  CENTRO OESTE         MATO GROSSO   
3        2004-05-09 2004-05-15  CENTRO OESTE  MATO GROSSO DO SUL   
4        2004-05-09 2004-05-15      NORDESTE             ALAGOAS   
...             ...        ...           ...                 ...   
120818   2021-04-25 2021-05-01         NORTE             RORAIMA   
120819   2021-04-25 2021-05-01           SUL      SANTA CATARINA   
120820   2021-04-25 2021-05-01       SUDESTE           SAO PAULO   
120821   2021-04-25 2021-05-01      NORDESTE             SERGIPE   
120822   2021-04-25 2021-05-01         NORTE           TOCANTINS   

                 PRODUTO  NÚMERO DE POSTOS PESQUISADOS UNIDADE DE MEDIDA  \
0       ETANOL HIDRATADO                           127              R$/l   
1       ETANOL HIDRATADO                           387              R$/l   
2       ETANOL HIDRATADO                           192              R$/l   
3       ETANOL HIDRATADO                           162              R$/l   
4       ETANOL HIDRATADO                           103              R$/l   
...                  ...                           ...               ...   
120818   OLEO DIESEL S10                            12              R$/l   
120819   OLEO DIESEL S10                           100              R$/l   
120820   OLEO DIESEL S10                           886              R$/l   
120821   OLEO DIESEL S10                            33              R$/l   
120822   OLEO DIESEL S10                            18              R$/l   

        PREÇO MÉDIO REVENDA  DESVIO PADRÃO REVENDA  PREÇO MÍNIMO REVENDA  \
0                     1.288                  0.016                 1.190   
1                     1.162                  0.114                 0.890   
2                     1.389                  0.097                 1.180   
3                     1.262                  0.070                 1.090   
4                     1.181                  0.078                 1.050   
...                     ...                    ...                   ...   
120818                4.582                  0.056                 4.510   
120819                4.141                  0.187                 3.838   
120820                4.246                  0.216                 3.769   
120821                4.464                  0.205                 4.190   
120822                4.352                  0.156                 4.199   

        PREÇO MÁXIMO REVENDA MARGEM MÉDIA REVENDA  COEF DE VARIAÇÃO REVENDA  \
0                      1.350                0.463                     0.012   
1                      1.449                0.399                     0.098   
2                      1.760                0.419                     0.070   
3                      1.509                0.432                     0.055   
4                      1.400                 0.24                     0.066   
...                      ...                  ...                       ...   
120818                 4.690             -99999.0                     0.012   
120819                 4.739             -99999.0                     0.045   
120820                 5.199             -99999.0                     0.051   
120821                 4.699             -99999.0                     0.046   
120822                 4.699             -99999.0                     0.036   

       PREÇO MÉDIO DISTRIBUIÇÃO DESVIO PADRÃO DISTRIBUIÇÃO  \
0                         0.825                       0.11   
1                         0.763                      0.088   
2                          0.97                      0.095   
3                          0.83                      0.119   
4                         0.941                      0.077   
...                         ...            

#  6. Utilizando o value_counts(), para listar todos os tipos de produtos contidos na base de dados.

In [19]:
df_combinado.value_counts("PRODUTO")

PRODUTO
GASOLINA COMUM        23570
GLP                   23561
ETANOL HIDRATADO      23440
ÓLEO DIESEL           21194
GNV                   14469
ÓLEO DIESEL S10        9113
OLEO DIESEL S10        2376
OLEO DIESEL            2351
GASOLINA ADITIVADA      749
Name: count, dtype: int64

#### 7. Filtrando o DataFrame para obter apenas dados da 'GASOLINA COMUM'. Gravando em uma nova variável.

In [20]:
gasolina_comum_df = df_combinado[df_combinado['PRODUTO'] == 'GASOLINA COMUM']
gasolina_comum_df


DATA INICIAL DATA FINAL        REGIÃO              ESTADO  \
12064    2004-05-09 2004-05-15  CENTRO OESTE    DISTRITO FEDERAL   
12065    2004-05-09 2004-05-15  CENTRO OESTE               GOIAS   
12066    2004-05-09 2004-05-15  CENTRO OESTE         MATO GROSSO   
12067    2004-05-09 2004-05-15  CENTRO OESTE  MATO GROSSO DO SUL   
12068    2004-05-09 2004-05-15      NORDESTE             ALAGOAS   
...             ...        ...           ...                 ...   
120721   2021-04-25 2021-05-01         NORTE             RORAIMA   
120722   2021-04-25 2021-05-01           SUL      SANTA CATARINA   
120723   2021-04-25 2021-05-01       SUDESTE           SAO PAULO   
120724   2021-04-25 2021-05-01      NORDESTE             SERGIPE   
120725   2021-04-25 2021-05-01         NORTE           TOCANTINS   

               PRODUTO  NÚMERO DE POSTOS PESQUISADOS UNIDADE DE MEDIDA  \
12064   GASOLINA COMUM                           128              R$/l   
12065   GASOLINA COMUM                           395              R$/l   
12066   GASOLINA COMUM                           194              R$/l   
12067   GASOLINA COMUM                           166              R$/l   
12068   GASOLINA COMUM                           106              R$/l   
...                ...                           ...               ...   
120721  GASOLINA COMUM                            12              R$/l   
120722  GASOLINA COMUM                           119              R$/l   
120723  GASOLINA COMUM                          1187              R$/l   
120724  GASOLINA COMUM                            34              R$/l   
120725  GASOLINA COMUM                            20              R$/l   

        PREÇO MÉDIO REVENDA  DESVIO PADRÃO REVENDA  PREÇO MÍNIMO REVENDA  \
12064                 2.029                  0.007                 1.990   
12065                 2.025                  0.062                 1.850   
12066                 2.358                  0.066                 2.000   
12067                 2.120                  0.075                 1.970   
12068                 2.090                  0.034                 2.000   
...                     ...                    ...                   ...   
120721                5.364                  0.065                 5.310   
120722                5.051                  0.251                 4.640   
120723                5.265                  0.218                 4.279   
120724                5.586                  0.088                 5.410   
120725                5.751                  0.160                 5.578   

        PREÇO MÁXIMO REVENDA MARGEM MÉDIA REVENDA  COEF DE VARIAÇÃO REVENDA  \
12064                  2.070                0.318                     0.003   
12065                  2.220                0.296                     0.031   
12066                  2.540                0.472                     0.028   
12067                  2.440                0.325                     0.035   
12068                  2.159                 0.35                     0.016   
...                      ...                  ...                       ...   
120721                 5.530             -99999.0                     0.012   
120722                 5.599             -99999.0                     0.050   
120723                 6.099             -99999.0                     0.041   
120724                 5.699             -99999.0                     0.016   
120725                 6.119             -99999.0                     0.028   

       PREÇO MÉDIO DISTRIBUIÇÃO DESVIO PADRÃO DISTRIBUIÇÃO  \
12064                     1.711                       0.02   
12065                     1.729                      0.036   
12066                     1.886                      0.068   
12067                     1.795                      0.033   
12068                      1.74                      0.042   
...                         ...                        ...   
12072

#### 8. Qual o preço médio de revenda da gasolina em agosto de 2008?

In [21]:
preco_medio_agosto_2008 = gasolina_comum_df[gasolina_comum_df['ANO-MES'] == '2008/08']['PREÇO MÉDIO REVENDA'].mean()
print(f"O preço médio de revenda da gasolina comum em agosto de 2008 foi de: R$ {preco_medio_agosto_2008:.3f}")

O preço médio de revenda da gasolina comum em agosto de 2008 foi de: R$ 2.601


# Esse valor está acima ou abaixo da média anual de 2008?

In [22]:
preco_medio_anual_2008 = gasolina_comum_df[gasolina_comum_df['DATA INICIAL'].dt.year == 2008]['PREÇO MÉDIO REVENDA'].mean()
preco_medio_agosto_2008 = gasolina_comum_df[gasolina_comum_df['ANO-MES'] == '2008/08']['PREÇO MÉDIO REVENDA'].mean()
diferenca = (preco_medio_agosto_2008 - preco_medio_anual_2008) / preco_medio_anual_2008 * 100
print(f"O preço médio de revenda da gasolina comum no ano de 2008 foi de: R$ {preco_medio_anual_2008:.3f}")
print(
    f"Em agosto de 2008, o preço médio da gasolina foi "
    f"{abs(diferenca):.3f}% {'abaixo' if diferenca < 0 else 'acima'} "
    f"da média anual."
)

O preço médio de revenda da gasolina comum no ano de 2008 foi de: R$ 2.605
Em agosto de 2008, o preço médio da gasolina foi 0.142% abaixo da média anual.


O preço médio de revenda da gasolina em agosto de 2008 apresentou valor muito próximo à média anual do mesmo ano, com variação inferior a 1%. Isso indica que agosto não se destacou como um mês atípico, sugerindo estabilidade dos preços nesse período.

#### 9. Qual o preço médio de revenda da gasolina em maio de 2014 em São Paulo?

In [23]:
preco_medio_maio_2014 = gasolina_comum_df[(gasolina_comum_df['ANO-MES'] == '2014/05') & (gasolina_comum_df['ESTADO'] == 'SAO PAULO')]['PREÇO MÉDIO REVENDA'].mean()
print(f"O preço médio de revenda da gasolina comum em maio de 2014 foi de: R$ {preco_medio_maio_2014:.3f}")

O preço médio de revenda da gasolina comum em maio de 2014 foi de: R$ 2.882


# Como esse valor se compara à média nacional no mesmo período?





In [24]:
preco_medio_nacional_maio_2014 = gasolina_comum_df[gasolina_comum_df['ANO-MES'] == '2014/05']['PREÇO MÉDIO REVENDA'].mean()
preco_medio_maio_2014 = gasolina_comum_df[(gasolina_comum_df['ANO-MES'] == '2014/05') & (gasolina_comum_df['ESTADO'] == 'SAO PAULO')]['PREÇO MÉDIO REVENDA'].mean()
diferenca = (preco_medio_maio_2014 - preco_medio_nacional_maio_2014) / preco_medio_nacional_maio_2014 * 100
print(f"O preço médio de revenda da gasolina comum nacional em maio de 2014 foi de: R$ {preco_medio_nacional_maio_2014:.3f}")
print(
    f"No mês de maio de 2014 em São Paulo, o preço médio da gasolina foi "
    f"{abs(diferenca):.3f}% {'abaixo' if diferenca < 0 else 'acima'} "
    f"da média nacional para o mesmo período."
)

O preço médio de revenda da gasolina comum nacional em maio de 2014 foi de: R$ 3.030
No mês de maio de 2014 em São Paulo, o preço médio da gasolina foi 4.892% abaixo da média nacional para o mesmo período.




# São Paulo estava acima ou abaixo da média do Sudeste?



In [25]:
preco_medio_sudeste_2014 = gasolina_comum_df[(gasolina_comum_df['ANO-MES'] == '2014/05') & (gasolina_comum_df['REGIÃO'] == 'SUDESTE')]['PREÇO MÉDIO REVENDA'].mean()
preco_medio_maio_2014 = gasolina_comum_df[(gasolina_comum_df['ANO-MES'] == '2014/05') & (gasolina_comum_df['ESTADO'] == 'SAO PAULO')]['PREÇO MÉDIO REVENDA'].mean()
diferenca = (preco_medio_maio_2014 - preco_medio_sudeste_2014) / preco_medio_sudeste_2014 * 100
print(f"O preço médio de revenda da gasolina comum no ano de 2014 foi de: R$ {preco_medio_sudeste_2014:.3f}")
print(
    f"No mês de maio de 2014 em São Paulo, o preço médio da gasolina foi "
    f"{abs(diferenca):.3f}% {'abaixo' if diferenca < 0 else 'acima'} "
    f"da média do sudeste no mesmo periodo."
)

O preço médio de revenda da gasolina comum no ano de 2014 foi de: R$ 2.999
No mês de maio de 2014 em São Paulo, o preço médio da gasolina foi 3.880% abaixo da média do sudeste no mesmo periodo.


Considerando que São Paulo possui um dos maiores volumes de consumo da região, o fato de apresentar preço médio aproximadamente 10% inferior à média do Sudeste indica uma posição competitiva relevante dentro da região Sudeste e também Nacional.

# Rio de janeiro teve a media semelhante com São Paulo em 2014 se comparar com Sudeste ?

In [26]:
preco_medio_rio_2014 = gasolina_comum_df[(gasolina_comum_df['ANO-MES'] == '2014/05') & (gasolina_comum_df['ESTADO'] == 'RIO DE JANEIRO')]['PREÇO MÉDIO REVENDA'].mean()
preco_medio_sudeste_2014 = gasolina_comum_df[(gasolina_comum_df['ANO-MES'] == '2014/05') & (gasolina_comum_df['REGIÃO'] == 'SUDESTE')]['PREÇO MÉDIO REVENDA'].mean()
diferenca = (preco_medio_rio_2014 - preco_medio_sudeste_2014) / preco_medio_sudeste_2014 * 100
print(
    f"No mês de maio de 2014 no Rio de Janeiro, o preço médio da gasolina foi "
    f"{abs(diferenca):.3f}% {'abaixo' if diferenca < 0 else 'acima'} "
    f"da média do sudeste no mesmo periodo."
)



No mês de maio de 2014 no Rio de Janeiro, o preço médio da gasolina foi 4.577% acima da média do sudeste no mesmo periodo.


Os resultados indicam assimetria de preços entre os estados, reforçando que a média regional não reflete igualmente a realidade de cada unidade federativa, enquanto São Paulo apresentou preços médios aproximadamente 4% abaixo da média de referência, o Rio de Janeiro apresentou valores cerca de 4,5% acima.Esse resultado sugere diferenças estruturais e competitivas entre os mercados estaduais.



# Esse mês foi atípico no histórico de preços de SP?

In [27]:
maios_sp = gasolina_comum_df[(gasolina_comum_df['ESTADO'] == 'SAO PAULO') & (gasolina_comum_df['DATA INICIAL'].dt.month == 5)]

media_historica_maio_sp = maios_sp['PREÇO MÉDIO REVENDA'].mean()

preco_medio_maio_2014_sp = gasolina_comum_df[(gasolina_comum_df['ANO-MES'] == '2014/05') &(gasolina_comum_df['ESTADO'] == 'SAO PAULO')]['PREÇO MÉDIO REVENDA'].mean()

diferenca_historica = (preco_medio_maio_2014_sp - media_historica_maio_sp) / media_historica_maio_sp * 100


print(f"O preço médio de revenda da gasolina comum em maio de 2014 em São Paulo foi de: R$ {preco_medio_maio_2014_sp:.3f}")
print(f"A média histórica de revenda da gasolina comum em maio para São Paulo é: R$ {media_historica_maio_sp:.3f}")
print(
    f"Em maio de 2014, o preço médio da gasolina em São Paulo foi "
    f"{abs(diferenca_historica):.3f}% {'abaixo' if diferenca_historica < 0 else 'acima'} "
    f"da média histórica de maio para o estado."
)

O preço médio de revenda da gasolina comum em maio de 2014 em São Paulo foi de: R$ 2.882
A média histórica de revenda da gasolina comum em maio para São Paulo é: R$ 2.904
Em maio de 2014, o preço médio da gasolina em São Paulo foi 0.737% abaixo da média histórica de maio para o estado.


In [28]:
maios_sp.head(10)

DATA INICIAL DATA FINAL   REGIÃO     ESTADO         PRODUTO  \
12087   2004-05-09 2004-05-15  SUDESTE  SAO PAULO  GASOLINA COMUM   
12114   2004-05-16 2004-05-22  SUDESTE  SAO PAULO  GASOLINA COMUM   
12141   2004-05-23 2004-05-29  SUDESTE  SAO PAULO  GASOLINA COMUM   
12168   2004-05-30 2004-06-05  SUDESTE  SAO PAULO  GASOLINA COMUM   
13464   2005-05-01 2005-05-07  SUDESTE  SAO PAULO  GASOLINA COMUM   
13491   2005-05-08 2005-05-14  SUDESTE  SAO PAULO  GASOLINA COMUM   
13518   2005-05-15 2005-05-21  SUDESTE  SAO PAULO  GASOLINA COMUM   
13545   2005-05-22 2005-05-28  SUDESTE  SAO PAULO  GASOLINA COMUM   
13572   2005-05-29 2005-06-04  SUDESTE  SAO PAULO  GASOLINA COMUM   
14840   2006-05-07 2006-05-13  SUDESTE  SAO PAULO  GASOLINA COMUM   

       NÚMERO DE POSTOS PESQUISADOS UNIDADE DE MEDIDA  PREÇO MÉDIO REVENDA  \
12087                          4107              R$/l                1.891   
12114                          4133              R$/l                1.888   
12141                          4139              R$/l                1.894   
12168                          4139              R$/l                1.912   
13464                          4104              R$/l                2.186   
13491                          4051              R$/l                2.186   
13518                          4105              R$/l                2.175   
13545                          4042              R$/l                2.167   
13572                          4094              R$/l                2.159   
14840                          3250              R$/l                2.476   

       DESVIO PADRÃO REVENDA  PREÇO MÍNIMO REVENDA  PREÇO MÁXIMO REVENDA  \
12087                  0.101                 1.570                 2.399   
12114                  0.101                 1.540                 2.220   
12141                  0.095                 1.570                 2.399   
12168                  0.094                 1.529                 2.199   
13464                  0.089                 1.929                 2.500   
13491                  0.087                 1.919                 2.699   
13518                  0.088                 1.919                 2.500   
13545                  0.090                 1.919                 2.699   
13572                  0.090                 1.919                 2.500   
14840                  0.106                 2.190                 2.990   

      MARGEM MÉDIA REVENDA  COEF DE VARIAÇÃO REVENDA PREÇO MÉDIO DISTRIBUIÇÃO  \
12087                0.235                     0.053                    1.656   
12114                0.233                     0.053                    1.655   
12141                0.234                     0.050                     1.66   
12168                0.235                     0.049                    1.677   
13464                0.249                     0.041                    1.937   
13491                0.256                     0.040                     1.93   
13518                0.256                     0.040                    1.919   
13545                0.256                     0.042                    1.911   
13572                0.255                     0.042                    1.904   
14840                0.263                     0.043                    2.213   

      DESVIO PADRÃO DISTRIBUIÇÃO PREÇO MÍNIMO DISTRIBUIÇÃO  \
12087                      0.064                      1.38   
12114                      0.063                      1.38   
12141                      0.065                      1.39   
12168                      0.068                      1.39   
13464                      0.041                      1.78   
13491                      0.045                      1.78   
13518                      0.047                     1.477   
13545                      0.046                    1.4869   
13572                      0.045                     1.477   
14840                      0.051      

In [29]:
preco_medio_maio_2014_sp = gasolina_comum_df[(gasolina_comum_df['ANO-MES'] == '2014/05') &(gasolina_comum_df['ESTADO'] == 'SAO PAULO')]
preco_medio_maio_2014_sp

DATA INICIAL DATA FINAL   REGIÃO     ESTADO         PRODUTO  \
66229   2014-04-27 2014-05-03  SUDESTE  SAO PAULO  GASOLINA COMUM   
66380   2014-05-04 2014-05-10  SUDESTE  SAO PAULO  GASOLINA COMUM   
66533   2014-05-11 2014-05-17  SUDESTE  SAO PAULO  GASOLINA COMUM   
66685   2014-05-18 2014-05-24  SUDESTE  SAO PAULO  GASOLINA COMUM   
66837   2014-05-25 2014-05-31  SUDESTE  SAO PAULO  GASOLINA COMUM   

       NÚMERO DE POSTOS PESQUISADOS UNIDADE DE MEDIDA  PREÇO MÉDIO REVENDA  \
66229                          2372              R$/l                2.888   
66380                          2373              R$/l                2.883   
66533                          2372              R$/l                2.886   
66685                          2370              R$/l                2.878   
66837                          2366              R$/l                2.876   

       DESVIO PADRÃO REVENDA  PREÇO MÍNIMO REVENDA  PREÇO MÁXIMO REVENDA  \
66229                  0.108                 2.490                 3.199   
66380                  0.110                 2.497                 3.219   
66533                  0.110                 2.100                 3.599   
66685                  0.109                 2.479                 3.699   
66837                  0.108                 2.569                 3.199   

      MARGEM MÉDIA REVENDA  COEF DE VARIAÇÃO REVENDA PREÇO MÉDIO DISTRIBUIÇÃO  \
66229                0.391                     0.037                    2.497   
66380                0.388                     0.038                    2.495   
66533                0.398                     0.038                    2.488   
66685                0.397                     0.038                    2.481   
66837                0.405                     0.038                    2.471   

      DESVIO PADRÃO DISTRIBUIÇÃO PREÇO MÍNIMO DISTRIBUIÇÃO  \
66229                      0.067                      2.27   
66380                      0.068                      2.27   
66533                      0.067                      2.29   
66685                      0.065                      2.29   
66837                      0.066                      2.29   

      PREÇO MÁXIMO DISTRIBUIÇÃO COEF DE VARIAÇÃO DISTRIBUIÇÃO  ANO-MES  
66229                     2.863                         0.027  2014/05  
66380                    2.7321                         0.027  2014/05  
66533                     2.855                         0.027  2014/05  
66685                    2.8245                         0.026  2014/05  
66837                    2.8245                         0.027  2014/05

 O preço médio da gasolina comum em São Paulo em maio de 2014 foi de R$ 2.882 , enquanto a média histórica foi de 2.904 Reais. Isso significa que, em maio de 2014, o preço médio da gasolina em São Paulo foi 0.737% abaixo da média histórica de maio para o estado. Essa variação percentual de menos de 1% em relação à média histórica indica que maio de 2014 não foi um mês atípico em termos de preço médio da gasolina em São Paulo. O preço se manteve muito próximo do que seria esperado historicamente para o mês de maio, sugerindo uma estabilidade relativa ou que fatores que influenciaram o preço naquele mês estavam em linha com as tendências históricas.

#### 10. Em qual(quais) estado(s) a gasolina ultrapassou a barreira dos R$ 5,00? E quando isso ocorreu?

In [30]:
gasolina_comum_df[gasolina_comum_df['PREÇO MÉDIO REVENDA'] >= 5] [["ESTADO", "ANO-MES", "PREÇO MÉDIO REVENDA"]].head(30)



ESTADO  ANO-MES  PREÇO MÉDIO REVENDA
98201             ACRE  2018/05                5.053
98347             ACRE  2018/06                5.035
98365   RIO DE JANEIRO  2018/06                5.000
98493             ACRE  2018/06                5.038
98511   RIO DE JANEIRO  2018/06                5.016
98646             ACRE  2018/06                5.031
98797             ACRE  2018/06                5.036
98948             ACRE  2018/06                5.039
99098             ACRE  2018/07                5.021
99247             ACRE  2018/07                5.042
99398             ACRE  2018/07                5.038
99549             ACRE  2018/07                5.033
99701             ACRE  2018/08                5.031
99852             ACRE  2018/08                5.015
100003            ACRE  2018/08                5.023
100153            ACRE  2018/08                5.036
100302            ACRE  2018/09                5.018
100453            ACRE  2018/09                5.074
100604            ACRE  2018/09                5.129
100756            ACRE  2018/09                5.153
100774  RIO DE JANEIRO  2018/09                5.036
100906            ACRE  2018/09                5.347
100924  RIO DE JANEIRO  2018/09                5.056
100932       TOCANTINS  2018/09                5.041
101055            ACRE  2018/10                5.345
101073  RIO DE JANEIRO  2018/10                5.070
101081       TOCANTINS  2018/10                5.049
101204            ACRE  2018/10                5.362
101222  RIO DE JANEIRO  2018/10                5.116
101230       TOCANTINS  2018/10                5.068

# Esses estados são também os mais voláteis?

In [31]:
preco_desvio_rio = gasolina_comum_df[gasolina_comum_df['ESTADO'] == 'RIO DE JANEIRO']["PREÇO MÉDIO REVENDA"].std()
preco_desvio_acre = gasolina_comum_df[gasolina_comum_df['ESTADO'] == 'ACRE']["PREÇO MÉDIO REVENDA"].std()
print(f"O desvio padrão para os Estados do Rio de Janeiro e Acre foram respectivamente: R${preco_desvio_rio:.3f}, R${preco_desvio_acre:.3f}")



O desvio padrão para os Estados do Rio de Janeiro e Acre foram respectivamente: R$0.936, R$0.837


O desvio padrão de aproximadamente R$ 0,90 indica elevada volatilidade nos preços, evidenciando oscilações significativas ao longo do período analisado.

# Houve propagação para outros estados logo depois?

In [32]:
gasolina_comum_df[(gasolina_comum_df['PREÇO MÉDIO REVENDA'] >= 5) & (gasolina_comum_df['DATA INICIAL'].dt.year == 2019)][["ESTADO", "ANO-MES", "PREÇO MÉDIO REVENDA"]]

ESTADO  ANO-MES  PREÇO MÉDIO REVENDA
105705            ACRE  2019/05                5.084
105855            ACRE  2019/05                5.088
105873  RIO DE JANEIRO  2019/05                5.004
106005            ACRE  2019/05                5.089
106156            ACRE  2019/06                5.121
106174  RIO DE JANEIRO  2019/06                5.044
106305            ACRE  2019/06                5.119
106455            ACRE  2019/06                5.076
106605            ACRE  2019/06                5.023
106754            ACRE  2019/06                5.012
106849            ACRE  2019/07                5.009

In [33]:
gasolina_comum_df[(gasolina_comum_df['PREÇO MÉDIO REVENDA'] >= 5) & (gasolina_comum_df['DATA INICIAL'].dt.year == 2020)][["ESTADO", "ANO-MES", "PREÇO MÉDIO REVENDA"]]

ESTADO  ANO-MES  PREÇO MÉDIO REVENDA
111060            ACRE  2020/01                5.024
111078  RIO DE JANEIRO  2020/01                5.043
111231  RIO DE JANEIRO  2020/01                5.106
111384  RIO DE JANEIRO  2020/02                5.025
111537  RIO DE JANEIRO  2020/02                5.019
111689  RIO DE JANEIRO  2020/02                5.025
111991  RIO DE JANEIRO  2020/02                5.007
112443  RIO DE JANEIRO  2020/03                5.012
116986            ACRE  2020/12                5.026
117160            ACRE  2020/12                5.051
117334            ACRE  2020/12                5.061
117506            ACRE  2020/12                5.063
117683            ACRE  2021/01                5.088

In [34]:
gasolina_comum_df[(gasolina_comum_df['PREÇO MÉDIO REVENDA'] >= 5) & (gasolina_comum_df['DATA INICIAL'].dt.year == 2021)][["ESTADO", "ANO-MES", "PREÇO MÉDIO REVENDA"]]

ESTADO  ANO-MES  PREÇO MÉDIO REVENDA
117857            ACRE  2021/01                5.068
118032            ACRE  2021/01                5.091
118208            ACRE  2021/01                5.213
118226  RIO DE JANEIRO  2021/01                5.015
118384            ACRE  2021/01                5.161
...                ...      ...                  ...
120721         RORAIMA  2021/05                5.364
120722  SANTA CATARINA  2021/05                5.051
120723       SAO PAULO  2021/05                5.265
120724         SERGIPE  2021/05                5.586
120725       TOCANTINS  2021/05                5.751

[285 rows x 3 columns]

Os primeiros preços acima de R$ 5,00 foram observados nos estados do Tocantins, Acre e Rio de Janeiro, expandindo-se para os demais estados principalmente em 2021, último ano analisado.

#### 11. Qual a média de preço dos estados da região sul em 2012?

In [36]:
media_sul = gasolina_comum_df[(gasolina_comum_df['REGIÃO'] == 'SUL') &(gasolina_comum_df['DATA INICIAL'].dt.year == 2012)]['PREÇO MÉDIO REVENDA'].mean()

print(f"O preço médio dos estados da região sul em 2012 foi de : R$ {media_sul:.3f}")

O preço médio dos estados da região sul em 2012 foi de : R$ 2.722


# Como a média da região Sul se compara às demais regiões em 2012?

In [43]:
media_sul = gasolina_comum_df[(gasolina_comum_df['REGIÃO'] == 'SUL') & (gasolina_comum_df['DATA INICIAL'].dt.year == 2012)]['PREÇO MÉDIO REVENDA'].mean()
media_sudeste = gasolina_comum_df[(gasolina_comum_df['REGIÃO'] == 'SUDESTE') & (gasolina_comum_df['DATA INICIAL'].dt.year == 2012)]['PREÇO MÉDIO REVENDA'].mean()
media_norte = gasolina_comum_df[(gasolina_comum_df['REGIÃO'] == 'NORTE') & (gasolina_comum_df['DATA INICIAL'].dt.year == 2012)]['PREÇO MÉDIO REVENDA'].mean()
media_nordeste = gasolina_comum_df[(gasolina_comum_df['REGIÃO'] == 'NORDESTE') & (gasolina_comum_df['DATA INICIAL'].dt.year == 2012)]['PREÇO MÉDIO REVENDA'].mean()
media_centro_oeste = gasolina_comum_df[(gasolina_comum_df['REGIÃO'] == 'CENTRO OESTE') & (gasolina_comum_df['DATA INICIAL'].dt.year == 2012)]['PREÇO MÉDIO REVENDA'].mean()

print(f"O preço médio dos estados da região sul em 2012 foi de : R$ {media_sul:.3f}")
print(f"O preço médio dos estados da região sudeste em 2012 foi de : R$ {media_sudeste:.3f}")
print(f"O preço médio dos estados da região norte em 2012 foi de : R$ {media_norte:.3f}")
print(f"O preço médio dos estados da região nordeste em 2012 foi de : R$ {media_nordeste:.3f}")
print(f"O preço médio dos estados da região centro oeste em 2012 foi de : R$ {media_centro_oeste:.3f}")



O preço médio dos estados da região sul em 2012 foi de : R$ 2.722
O preço médio dos estados da região sudeste em 2012 foi de : R$ 2.784
O preço médio dos estados da região norte em 2012 foi de : R$ 2.900
O preço médio dos estados da região nordeste em 2012 foi de : R$ 2.689
O preço médio dos estados da região centro oeste em 2012 foi de : R$ 2.838


In [46]:
diferenca_sudeste = (media_sul - media_sudeste) / media_sudeste * 100
diferenca_norte = (media_sul - media_norte) / media_norte * 100
diferenca_nordeste = (media_sul - media_nordeste) / media_nordeste * 100
diferenca_centro_oeste = (media_sul - media_centro_oeste) / media_centro_oeste * 100

print(f"A diferença dos estados da região sul em relação à região sudeste foi de "f"{abs(diferenca_sudeste):.3f}% {'abaixo' if diferenca_sudeste < 0 else 'acima'} ")
print(f"A diferença dos estados da região sul em relação à região norte foi de: "f"{abs(diferenca_norte):.3f}% {'abaixo' if diferenca_norte < 0 else 'acima'} ")
print(f"A diferença dos estados da região sul em relação à região nordeste foi de: "f"{abs(diferenca_nordeste):.3f}% {'abaixo' if diferenca_nordeste < 0 else 'acima'} ")
print(f"A diferença dos estados da região sul em relação à região centro oeste foi de: "f"{abs(diferenca_centro_oeste):.3f}% {'abaixo' if diferenca_centro_oeste < 0 else 'acima'} ")

A diferença dos estados da região sul em relação à região sudeste foi de 2.197% abaixo 
A diferença dos estados da região sul em relação à região norte foi de: 6.109% abaixo 
A diferença dos estados da região sul em relação à região nordeste foi de: 1.243% acima 
A diferença dos estados da região sul em relação à região centro oeste foi de: 4.087% abaixo 


#### 12. Uma tabela contendo a variação percentual ano a ano para o estado do Rio de Janeiro

In [47]:
dados_rio_de_janeiro = gasolina_comum_df[gasolina_comum_df['ESTADO'] == 'RIO DE JANEIRO'].copy()

preco_medio_anual_rj = dados_rio_de_janeiro.groupby(dados_rio_de_janeiro['DATA INICIAL'].dt.year)['PREÇO MÉDIO REVENDA'].mean()

variacao_percentual_anual = preco_medio_anual_rj.pct_change() * 100

resultado_rj = pd.DataFrame({'Preço Médio Anual (R$)': preco_medio_anual_rj,'Variação Percentual Anual (%)': variacao_percentual_anual})

print("Variação percentual ano a ano do preço médio da gasolina no Rio de Janeiro:")
display(resultado_rj)

Variação percentual ano a ano do preço médio da gasolina no Rio de Janeiro:


Preço Médio Anual (R$)  Variação Percentual Anual (%)
DATA INICIAL                                                       
2004                        2.175500                            NaN
2005                        2.339500                       7.538497
2006                        2.560679                       9.454125
2007                        2.532288                      -1.108721
2008                        2.547308                       0.593109
2009                        2.566120                       0.738517
2010                        2.648673                       3.217039
2011                        2.835846                       7.066673
2012                        2.853698                       0.629511
2013                        2.998885                       5.087662
2014                        3.134635                       4.526683
2015                        3.549882                      13.247086
2016                        3.919192                      10.403442
2017                        4.112774                       4.939316
2018                        4.858058                      18.121204
2019                        4.858173                       0.002375
2020                        4.761205                      -1.995988
2021                        5.625294                      18.148550

# Quais anos tiveram maior variação positiva?

Observam-se variações positivas mais acentuadas em 2015, 2016, 2018 e 2021, indicando períodos de aumento mais intenso dos preços.

# O RJ apresenta maior volatilidade que a média nacional?

In [51]:
dados_nacional = gasolina_comum_df

preco_medio_anual = dados_nacional.groupby(dados_nacional['DATA INICIAL'].dt.year)['PREÇO MÉDIO REVENDA'].mean()

variacao_percentual = preco_medio_anual.pct_change() * 100

resultado = pd.DataFrame({'Preço Médio Anual (R$)': preco_medio_anual,'Variação Percentual Anual (%)': variacao_percentual})

print("Variação percentual ano a ano do preço médio da gasolina nacional:")
display(resultado)

std_rj = variacao_percentual_anual.std()
std_nacional = variacao_percentual.std()

print(f"\nDesvio padrão da variação percentual anual no Rio de Janeiro: {std_rj:.2f}%")
print(f"Desvio padrão da variação percentual anual nacional: {std_nacional:.2f}%")

if std_rj > std_nacional:
    print("O Rio de Janeiro apresenta maior volatilidade que a média nacional.")
elif std_rj < std_nacional:
    print("O Rio de Janeiro apresenta menor volatilidade que a média nacional.")
else:
    print("O Rio de Janeiro apresenta volatilidade semelhante à média nacional.")

Variação percentual ano a ano do preço médio da gasolina nacional:


Preço Médio Anual (R$)  Variação Percentual Anual (%)
DATA INICIAL                                                       
2004                        2.238857                            NaN
2005                        2.436973                       8.848979
2006                        2.654691                       8.933943
2007                        2.602651                      -1.960308
2008                        2.604931                       0.087600
2009                        2.612350                       0.284824
2010                        2.661623                       1.886150
2011                        2.781453                       4.502131
2012                        2.783480                       0.072879
2013                        2.915976                       4.760074
2014                        3.034759                       4.073541
2015                        3.421235                      12.734958
2016                        3.748847                       9.575850
2017                        3.814546                       1.752510
2018                        4.444822                      16.522968
2019                        4.412135                      -0.735386
2020                        4.299953                      -2.542589
2021                        5.183976                      20.558904


Desvio padrão da variação percentual anual no Rio de Janeiro: 6.27%
Desvio padrão da variação percentual anual nacional: 6.69%
O Rio de Janeiro apresenta menor volatilidade que a média nacional.


# Existe tendência de alta consistente ao longo dos anos?

Observa-se que as quedas apresentam comportamento relativamente consistente, muitas vezes próximas de 0%, enquanto as variações positivas tendem a se intensificar ao longo dos anos, passando de valores médios em torno de 10% para patamares próximos de 18%.

#### Criando uma tabela contendo uma serie temporal com a diferença absoluta e percentual entre os valores mais baratos e caros.Também ao lado os estados na qual os maiores e menores preços foram registrados.

In [52]:
# Agrupar por 'ANO-MES' para encontrar os preços mínimo e máximo e os estados correspondentes
min_max_prices = gasolina_comum_df.groupby('ANO-MES').agg(
    min_preco=('PREÇO MÉDIO REVENDA', 'min'),
    max_preco=('PREÇO MÉDIO REVENDA', 'max')
).reset_index()

# Função para obter o estado com o preço mínimo/máximo
def get_state_for_price(df, year_month, price_col, price_value):
    temp_df = df[(df['ANO-MES'] == year_month) & (df[price_col] == price_value)]
    if not temp_df.empty:
        return temp_df['ESTADO'].iloc[0] # Retorna o primeiro estado encontrado
    return None

# Adicionar os estados correspondentes aos preços mínimo e máximo
min_max_prices['estado_menor_preco'] = min_max_prices.apply(lambda row: get_state_for_price(gasolina_comum_df, row['ANO-MES'], 'PREÇO MÉDIO REVENDA', row['min_preco']), axis=1)
min_max_prices['estado_maior_preco'] = min_max_prices.apply(lambda row: get_state_for_price(gasolina_comum_df, row['ANO-MES'], 'PREÇO MÉDIO REVENDA', row['max_preco']), axis=1)

# Calcular a diferença absoluta e percentual
min_max_prices['diferenca_absoluta'] = min_max_prices['max_preco'] - min_max_prices['min_preco']
min_max_prices['diferenca_percentual'] = (min_max_prices['diferenca_absoluta'] / min_max_prices['min_preco']) * 100

# Exibir a tabela resultante
print("Tabela de Variação de Preços (Mínimo vs. Máximo) por Ano-Mês:")
display(min_max_prices.head())


Tabela de Variação de Preços (Mínimo vs. Máximo) por Ano-Mês:


ANO-MES  min_preco  max_preco estado_menor_preco estado_maior_preco  \
0  2004/05      1.888      2.438          SAO PAULO               ACRE   
1  2004/06      1.912      2.496          SAO PAULO        MATO GROSSO   
2  2004/07      1.936      2.545   DISTRITO FEDERAL               ACRE   
3  2004/08      2.030      2.560          SAO PAULO        MATO GROSSO   
4  2004/09      2.031      2.570          SAO PAULO               ACRE   

   diferenca_absoluta  diferenca_percentual  
0               0.550             29.131356  
1               0.584             30.543933  
2               0.609             31.456612  
3               0.530             26.108374  
4               0.539             26.538651